python/tank_vendor/shotgun_deploy/descriptor_io/dev.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+# 
+# CONFIDENTIAL AND PROPRIETARY
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+from .path import IODescriptorPath
+from ..errors import ShotgunDeployError
+
+from .. import util
+log = util.get_shotgun_deploy_logger()
+
+class IODescriptorDev(IODescriptorPath):
+    """
+    Represents a local dev item. This item is never downloaded
+    into the local storage, you interact with it directly::
+
+        {"type": "dev", "path": "/path/to/app"}
+
+    Optional parameters are possible::
+
+        {"type": "dev", "path": "/path/to/app", "name": "my-app"}
+
+        {"type": "dev",
+         "linux_path": "/path/to/app",
+         "windows_path": "d:\foo\bar",
+         "mac_path": "/path/to/app" }
+
+    String urls are on the following form::
+
+        sgtk:dev:[name]:local_path
+        sgtk:dev3:[name]:win_path:linux_path:mac_path
+
+        sgtk:dev:my-app:/tmp/foo/bar
+        sgtk:dev3::c%3A%0Coo%08ar:/tmp/foo/bar:
+
+    Name is optional and if not specified will be determined based on folder path.
+    If name is not specified and path is /tmp/foo/bar, the name will set to 'bar'
+    """
+
+    def __init__(self, location_dict):
+        """
+        Constructor
+
+        :param location_dict: Location dictionary describing the bundle
+        :return: Descriptor instance
+        """
+        super(IODescriptorDev, self).__init__(location_dict)
+
+    @classmethod
+    def dict_from_uri(cls, uri):
+        """
+        Given a location uri, return a location dict
+
+        :param uri: Location uri string
+        :return: Location dictionary
+        """
+        # sgtk:dev:[name]:local_path
+        # sgtk:dev3:[name]:win_path:linux_path:mac_path
+        #
+        # Examples:
+        # sgtk:dev:my-app:/tmp/foo/bar
+        # sgtk:dev3::c%3A%0Coo%08ar:/tmp/foo/bar:
+
+        # explode into dictionary
+        location_dict = None
+        try:
+            location_dict = cls._explode_uri(
+                uri,
+                "dev",
+                ["name", "path"]
+            )
+        except ShotgunDeployError:
+            # probably because it's a dev3, not a dev
+            pass
+
+        try:
+            location_dict = cls._explode_uri(
+                uri,
+                "dev3",
+                ["name", "windows_path", "linux_path", "mac_path"]
```

Comment:
random thoughts... feel free to ignore: 

The new uri stuff is really cool. I like that you can identify a location via string. I had a couple of thoughts/questions just for the sake of conversation...

1. Crazy thought... Would it make any sense to leverage the templating system to create a spec for each uri type? Maybe the templating system is strictly path based, but I was wondering if there would be any benefit either in terms of customization or future proofing to externalize the location uris specs: 
```yaml
def_template: "sgtk:dev:{name}:{path}"
```
2. Assuming 1 is just ridiculous, I'm curious why not use a standard URI format with a custom scheme such as `sgtk:///my-app?path=/tmp/foo/bar`. That's not very pretty, but would allow for standard parsing, encoding, optional params, etc. Perhaps other benefits down the road somewhere?